### PR TITLE
Fix edge cases for long and short decimal formatters

### LINKED ISSUE #91: LongDecimalFormatter uses wrong word when rounding up
**Describe the bug**

`LongDecimalFormatter` uses the wrong word when rounding up. For example, there's no "1000 thousand" in English or "10千" in Chinese:

```
999 thousand, 1000 thousand, 1 million
9千, 10千, 1万
```

**To Reproduce**

```js
console.log([
	['en', 999_499, 999_500, 1_000_000],
	['zh-cn', 9_499, 9_500, 10_000],
].map(([locale, ...ns]) => {
	const TwitterCldr = require('twitter_cldr').load(locale)

	return ns.map((n) => {
		return new TwitterCldr.LongDecimalFormatter().format(n)
	}).join(', ')
}).join('\n'))
```

**Expected behavior**

Output for 999,500 in `en` should be "1 million". Output for 9,500 in `zh-cn` should be "1万".

**Environment**

twitter_cldr 2.2.0, Node v16.13.0

### PR DESCRIPTION
**Pull request by lionel-rowe**
Fixes https://github.com/twitter/twitter-cldr-js/issues/91.

For input:
```js
[
	['en', 999_499, 999_500, 1_000_000],
	['zh-cn', 9_499, 9_500, 10_000],
]
```

Before:
```
999 thousand, 1000 thousand, 1 million
9千, 10千, 1万
```

After:
```
999 thousand, 1 million, 1 million
9千, 1万, 1万
```

More examples in `spec/js/numbers/long_short_decimal.spec.js`.

### CODE REVIEW

File: lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee [hunk 1]
```diff
@@ -21,30 +21,47 @@ class TwitterCldr.NumberFormatter
...
-      sign = if number < 0 && prefix != "-" then @symbols().minus_sign || @default_symbols.minus_sign else ""
+      sign = if number_part < 0 && prefix != "-" then @symbols().minus_sign || @default_symbols.minus_sign else ""
       "#{prefix}#{result}#{suffix}"
     else
-      # there's no specific formatting pattern for this number in current locale
-      number.toString()
+      unless @ instanceof TwitterCldr.DecimalFormatter
+        # No specific formatting pattern for this number in current locale, so
+        # we fall back to DecimalFormatter
+        new TwitterCldr.DecimalFormatter().format(number, opts)
```
Comment (lionel-rowe) [inline lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee:61]: Tbh I don't love instantiating the derived class from the base class, but for derived classes other than `DecimalFormatter`, it seems like a more sensible default than just calling `toString` (which completely ignores the correct formatting for that locale and just formats using `.` as a decimal point). Maybe there's a less kludgey way of refactoring this, though?

File: lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee [hunk 2]
```diff
@@ -106,7 +123,7 @@ class TwitterCldr.DecimalFormatter extends TwitterCldr.NumberFormatter
           result = transliterator.transliterate(result)
       result
     catch error
-      number
+      number.toString()
```
Comment (lionel-rowe) [inline lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee:126]: Previously was just returning the number itself, making the return type `string | number` instead of just `string`

File: spec/js/numbers/long_short_decimal.spec.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  Long: {
+    "en": [
+      // edge cases
+      { raw: 999_499, precisions: ["999 thousand", "999.5 thousand"] },
+      { raw: 999_500, precisions: ["1 million", "999.5 thousand"] },
+      { raw: 1_000_000, precisions: ["1 million", "1.0 million"] },
+
+      // fractional
+      { raw: 999.49, precisions: ["999", "999.5"], default: "999.49" },
+      { raw: 999.5, precisions: ["1 thousand", "999.5"], default: "999.5" },
```
Comment (lionel-rowe) [inline spec/js/numbers/long_short_decimal.spec.js:16]: I did originally try adding some test cases like `1_000_000.5`, but tbh it's not even clear to me what that should be formatted to at `precision = 1` — `1,000,000.5`? `1.0 million`? Something else entirely? Plus it's enough of a corner case that it's probably not worth worrying about as long as it does something vaguely sane.

File: lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee (hunk 1 above)
Comment (camertron) [reply lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee:61]: Is this necessary as a fallback in case there's no formatting data for the language/number combo?

File: spec/js/numbers/long_short_decimal.spec.js (hunk 3 above)
Comment (camertron) [reply spec/js/numbers/long_short_decimal.spec.js:13]: Generally speaking these formatters are used to format really large numbers so the decimal part isn't significant. Both twitter-cldr-rb and ICU format `1_000_000.5` (precision 1) as "1.0 million". However twitter-cldr-rb outputs "1.0000005 million" when formatted with a precision of 7, which means the decimal isn't entirely discarded. In any case I think these tests are legit.

**Review by camertron — APPROVED** [review]

File: lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee (hunk 1 above)
Comment (lionel-rowe) [reply lib/twitter_cldr/js/mustache/implementation/numbers/numbers.coffee:61]: TBH I've already forgotten where this was necessary — it seems the parent `else` code path should never actually be hit with any of the current `NumberFormatter`-derived classes, as `tokens` should always be defined. In which case could remove the `unless` and keep the fallback as `toString`, with a comment noting that it's only to prevent errors in case of bugs?

**Comment by camertron** [comment]
Apologies this took so long, but this has been published in v3.3.0.
